### ISSUE TITLE
Support for SamAutomaticMaskGenerator

### ISSUE THREAD

**Issue by petebankhead** [ksugar/samapi#5]
Hi,

Would you consider adding `SamAutomaticMaskGenerator` support? If so, what would be the best way to add it?

I think this would be a really nice addition, since it could then be incorporated into the [QuPath SAM extension](https://github.com/ksugar/qupath-extension-sam).

*Ideally*, the QuPath extension could use `SamAutomaticMaskGenerator` both interactively *and* from a script because it doesn't inherently require user interaction, and it would be good if the various `kwargs` could be supported as well.

From the QuPath side, I imagine that a user might call a script that looks like this:
```groovy
SAMUtils.segmentEverything("vit-l", ['points_per_side': 32, 'pred_iou_thresh': 0.86])
```
which would operate on the current image, or even current annotation selected within an image if required.

With the help of QuPath's 'Run for project', it would be possible to generate an initial segmentation across large numbers of images quickly... and, from Groovy, filter out segmented objects with specific characteristics if needed (e.g. very high or low mean intensity values).

**Comment by ksugar** [comment]
Hi @petebankhead , thank you for the suggestion.
I have written a prototype of it. https://github.com/ksugar/samapi/commit/294a23372eee3e16869cd6774eba661f40a072b3

![qupath-segment-everything-1080p](https://github.com/ksugar/samapi/assets/36190671/285a7eeb-2553-4105-b807-3a049047473f)

Currently, the server just takes a base64-encoded image, leaving the choice of an input image to the client side.
Does this achieve what you have in mind?
It is a good idea to cache the image data as you did here.
https://github.com/petebankhead/samapi/blob/d08d255bbe3283df18d6bb3efeac803fb64ab576/src/samapi/main.py#L104-L109

The client will be adapted to the latest commit: https://github.com/ksugar/qupath-extension-sam/pull/3.

**Comment by petebankhead** [comment]
Hi @ksugar looks good! Yes, that's pretty much what I was thinking. If caching the image, I think you will need to retain the predictor since  the `SamPredictor` constructor calls `reset_image` [here](https://github.com/facebookresearch/segment-anything/blob/6fdee8f2727f4506cfbbe553e23b895e27956588/segment_anything/predictor.py#L17).

Within the client, one possible way to make the functionality available would be via providing a different [`SAMOutputType`](https://github.com/petebankhead/qupath-extension-sam/blob/proposed/src/main/java/org/elephant/sam/SAMOutput.java), although it would probably be better to use a separate dialog.

**Comment by ksugar** [comment]
It's released with [v0.3.0](https://github.com/ksugar/samapi/releases/tag/v0.3.0).

@petebankhead I have made some modifications to SamAutomaticMaskGenerator to enable the cache of the encoder outputs.
Please see the details in the following commit.
https://github.com/ksugar/MobileSAM/commit/c6a2bab8e168f7745763f7704d44449156142d9d
